[Woo POS] Animated transition to the Get Support screen

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12628
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR attempts to animate the transition to the Get Support screen as per design feedback and animations specs (p6riRB-aNq-p2).

<img width="1311" alt="Screenshot 2024-09-17 at 12 44 58" src="https://github.com/user-attachments/assets/8fdf03c2-12ca-48c7-ac63-0f81424aabf1">

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
1. Enter POS mode
2. Tap "Get Support" from the bottom menu
3. Observe transition to the support screen

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
Transitions should follow animation specs (see above), and animate by sliding horizontally between screens.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
As shown in the section below, I checked the animations of transitions when entering and exiting the "Get Support" screen. In the store management mode, the transitions should remain unchanged.

Devices: Galaxy Tab S9 FE, Android 14

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
#### Before
https://github.com/user-attachments/assets/593f4cd2-c295-44f0-b3e7-d764713f3e04

#### After
https://github.com/user-attachments/assets/bca0190d-5998-415b-b700-aa4494eb2c21

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Implement transition to the Get Support screen

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Merge branch 'trunk' into 12628-transition-to-get-support-screen

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/support/requests/SupportRequestFormActivity.kt:
```diff
@@ -55,6 +57,40 @@ class SupportRequestFormActivity : AppCompatActivity() {
+
+    @Suppress("DEPRECATION")
+    private fun isPOS(): Boolean {
+        val origin = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
```
@samiuelson 👋

Take a look into `DataContainerExt.kt` file - I think we have extensions there

samiuelson replied:
Updated — [d7548d1](https://github.com/woocommerce/woocommerce-android/pull/12630/commits/d7548d112609642d31726fbaa586c85e7ef1f2c6)

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/support/requests/SupportRequestFormActivity.kt:
```diff
@@ -55,6 +57,40 @@ class SupportRequestFormActivity : AppCompatActivity() {
+    @Suppress("DEPRECATION")
+    private fun SupportRequestFormActivity.adjustActivityTransitions() {
+        if (isPOS()) {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
```
The same `SystemVersionUtils`. Not sure though if we want to reuse both in the POS

samiuelson replied:
Done — [7a8b219](https://github.com/woocommerce/woocommerce-android/pull/12630/commits/7a8b219f0ddde321e61043f37f44f36bf69a0a74)

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/support/requests/SupportRequestFormActivity.kt:
```diff
@@ -55,6 +57,40 @@ class SupportRequestFormActivity : AppCompatActivity() {
+    private fun SupportRequestFormActivity.adjustActivityTransitions() {
+        if (isPOS()) {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
+                overrideActivityTransition(
```
Is this purposefully that we we kinda going back (from left to right?). Shouldn't it be from right to left to show that we go forward here? If that's true, then there

```
android.R.anim.slide_in_left
android.R.anim.slide_out_right
```

to be reused

backwardstruck replied:
That seems to be a reasonable conclusion. I don't think it's actually specified in the design (unless I didn't see it).

samiuelson replied:
Done — [afc7e26](https://github.com/woocommerce/woocommerce-android/pull/12630/commits/afc7e26cfcdd8a5f842ca338985d26c0b10bd54c) and [4a790bb](https://github.com/woocommerce/woocommerce-android/pull/12630/commits/4a790bbd76a28514a86bdc7e60abd1fbf0a48896)

backwardstruck committed:
Merge branch 'trunk' into 12628-transition-to-get-support-screen

backwardstruck commented:
I think this is the issue you meant to link:
https://github.com/woocommerce/woocommerce-android/issues/12628

samiuelson committed:
Update transition animation specs

- Reuse android ones
- Update duration

samiuelson committed:
Simplify isPOS() fun

samiuelson committed:
Change animation to slide-in right

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/237">20.5</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Simplify the code

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Merge branch 'trunk' into 12628-transition-to-get-support-screen

samiuelson commented:
Thanks for review, @kidinov! I updated the animations:

https://github.com/user-attachments/assets/4ef35431-86ec-4638-967a-495ead9b3421

Let me know what you think.

kidinov commented:
@samiuelson, my understanding is that we should have something like this (no alpha, just moving from the right to left and backwards). The animations from android do have alpha so custom are needed

Here is the patch for android 34+
[Animation_for_34+.patch](https://github.com/user-attachments/files/17046084/Animation_for_34%2B.patch)

https://github.com/user-attachments/assets/36c633c1-99a0-42a0-b3ae-9beb404dc255

For Android <34, we also need to add `overridePendingTransition` in the activity's finish method to have proper back animation

backwardstruck committed:
Merge branch 'trunk' into 12628-transition-to-get-support-screen

samiuelson committed:
Update animation specs

samiuelson committed:
Fix transition on OS level < 34

samiuelson committed:
Merge remote-tracking branch 'origin/12628-transition-to-get-support-screen' into 12628-transition-to-get-support-screen

samiuelson committed:
Fix condition for transition on OS level < 34

samiuelson committed:
Fix transition implementation on OS level < 34

samiuelson commented:
> @samiuelson, my understanding is that we should have something like this (no alpha, just moving from the right to left and backwards). The animations from android do have alpha so custom are needed

Makes sense, updated — [278c0b8](https://github.com/woocommerce/woocommerce-android/pull/12630/commits/278c0b8d0b2828b08494b07da80523cf798aead0)

> For Android <34, we also need to add `overridePendingTransition` in the activity's finish method to have proper back animation

Good catch, fixed:

https://github.com/user-attachments/assets/7541aee2-8468-40c8-9aa7-08e7528198ff

Let me know what you think, @kidinov.

samiuelson committed:
Merge branch 'trunk' into 12628-transition-to-get-support-screen

samiuelson committed:
Clean up code

samiuelson committed:
Merge remote-tracking branch 'origin/12628-transition-to-get-support-screen' into 12628-transition-to-get-support-screen

samiuelson commented:
👋🏼 @backwardstruck, @kidinov, this is ready for review again.

kidinov reviewed (commented):
> * Can we just override `finish` method instead of calling `adjustExitTransition` from multiples places? I haven't tried it, but I recall that was the way
> 
> ```
>     override fun finish() {
>         adjustExitTransition()
>         super.finish()
>     }
> ```

Good catch, thanks @kidinov ! I think it should be called right after `super.finish()`; Fixed – [a1fe6af](https://github.com/woocommerce/woocommerce-android/pull/12630/commits/a1fe6afa2df4007ae86270112218595c43b1afad)

> I think on Android <34 it slides in the opposite direction it should (from left to right)

Fixed – [efb7e35](https://github.com/woocommerce/woocommerce-android/pull/12630/commits/efb7e3519701be1c1956099db9858a1ef2c80ef1)

backwardstruck committed:
Merge branch 'trunk' into 12628-transition-to-get-support-screen

samiuelson committed:
Clean up code by overriding finish()

samiuelson committed:
Fix animation sliding directions

samiuelson committed:
Merge remote-tracking branch 'origin/12628-transition-to-get-support-screen' into 12628-transition-to-get-support-screen

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/238">20.6</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Merge branch 'trunk' into 12628-transition-to-get-support-screen

kidinov committed:
Removed unused import

kidinov reviewed (approved):
Looks great!

I removed unused import to make it mergable 

a713b1ae17064ba169c47a28bf5c45b368232914
